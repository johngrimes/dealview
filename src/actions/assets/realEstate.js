// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import AssetActions from 'actions/assets/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import { createObjectActions } from 'actions/objects'
import type { RealEstate } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'

const actions = createObjectActions('REAL_ESTATE', 'REAL_ESTATE',
                                    'RealEstate', 'RealEstate')

const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(actions.putRealEstateRequest(realEstate))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, realEstate)
        dispatch(actions.putRealEstateSuccess(saved))
        await dispatch(AssetActions.putAsset(realEstateToAsset(saved)))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteRealEstateRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const deletedId = await deleteObject(objectStore, id)
        dispatch(AssetActions.deleteAsset(deletedId))
        dispatch(actions.deleteRealEstateSuccess(deletedId))
        resolve(deletedId)
      } catch (error) {
        dispatch(actions.deleteRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(actions.loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(actions.loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export default {
  ...actions,
  putRealEstate,
  deleteRealEstate,
  loadRealEstate,
}
