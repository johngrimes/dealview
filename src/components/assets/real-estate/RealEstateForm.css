.real-estate-form {
  padding: 0 16px;
}

.form-page-inactive {
  display: none;
}
