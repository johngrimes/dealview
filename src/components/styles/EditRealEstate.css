.edit-real-estate {
  padding: 0 16px;
}

.edit-real-estate .breadcrumbs {
  margin: 0 -16px 16px;
}
