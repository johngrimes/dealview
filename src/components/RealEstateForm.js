import React from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import moment from 'moment'

import InputField from './InputField.js'
import CurrencyField from './CurrencyField.js'
import HiddenField from './HiddenField.js'
import AddressField, { AddressErrorsDefaults } from './AddressField.js'
import TextAreaField from './TextAreaField.js'
import ValuationsInput from './ValuationsInput.js'
import DateField from './DateField.js'
import * as Validations from '../utils/formValidation.js'
import { RealEstateDefaults } from '../data/realEstate.js'
import { DateStorageFormat } from '../data/commonTypes.js'
import {
  updateValuationsWithPurchaseOrSale,
  updateValuationsWithPurchase,
  updateValuationsWithSale,
  purchaseFilter,
  saleFilter,
} from '../data/valuations.js'

import '../styles/forms.css'
import '../styles/buttons.css'
import '../styles/tables.css'
import './styles/RealEstateForm.css'

const RealEstateErrorsDefaults = {
  name: [],
  address: AddressErrorsDefaults,
  notes: [],
  purchaseDate: [],
  purchaseAmount: [],
  saleDate: [],
  saleAmount: [],
  valuations: [],
}

class RealEstateForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      realEstate: RealEstateDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.realEstate) {
      const realEstate = props.realEstate
      const valuations = props.realEstate.valuations
        ? props.realEstate.valuations
        : []
      this.state = {
        ...this.state,
        realEstate,
        purchase: _.pick(valuations.find(purchaseFilter), [ 'date', 'amount' ]),
        sale: _.pick(valuations.find(saleFilter), [ 'date', 'amount' ]),
        errors: RealEstateForm.validate(realEstate),
        allErrorsShown: true,
      }
    }

    this.handleFocus = this.handleFocus.bind(this)
    this.handleAddressChange = this.handleAddressChange.bind(this)
    this.handleValuationsChange = this.handleValuationsChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName, value) {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, [fieldName]: value }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleFocus(fieldName) {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleAddressChange(address) {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, address }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleValuationsChange(valuations) {
    this.setState(prevState => {
      let purchaseVal = valuations.find(v => v.type === 'purchase')
      let saleVal = valuations.find(v => v.type === 'sale')
      let updatedValuations = valuations
      if (purchaseVal) {
        purchaseVal = _.pick(purchaseVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithPurchase(
          valuations,
          purchaseVal
        )
      }
      if (saleVal) {
        saleVal = _.pick(saleVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithSale(valuations, saleVal)
      }
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
      }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseOrSaleChange(purchaseOrSale, type) {
    this.setState(prevState => {
      const updatedPurchaseOrSale = { ...prevState[type], ...purchaseOrSale }
      const updatedValuations = updateValuationsWithPurchaseOrSale(
        prevState.realEstate.valuations,
        updatedPurchaseOrSale,
        type
      )
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
        [type === 'purchase' // eslint-disable-line
          ? 'purchaseDate'
          : 'saleDate']: updatedPurchaseOrSale.date,
      }
      return {
        ...prevState,
        realEstate,
        [type === 'purchase' ? 'purchase' : 'sale']: updatedPurchaseOrSale,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseChange(purchase) {
    return this.handlePurchaseOrSaleChange(purchase, 'purchase')
  }

  handleSaleChange(sale) {
    return this.handlePurchaseOrSaleChange(sale, 'sale')
  }

  handleSubmit(event) {
    event.preventDefault()
    this.setState(
      () => ({ errors: RealEstateForm.validate(this.state.realEstate) }),
      () => {
        const { errors, realEstate } = this.state
        if (
          typeof errors !== 'undefined' &&
          Validations.areErrorsPresent(errors)
        ) {
          const firstErrorFieldName = Validations.findFirstErrorFieldName(
            errors
          )
          firstErrorFieldName
            ? this.setState(() => ({
              allErrorsShown: true,
              focusedInput: firstErrorFieldName,
            }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) {
            this.props.onSubmit(realEstate)
          }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.realEstate) {
      this.setState(() => ({ realEstate: nextProps.realEstate }))
    }
  }

  render() {
    const {
      realEstate,
      purchase,
      sale,
      allErrorsShown,
      focusedInput,
    } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField =
      typeof realEstate.id === 'string'
        ? <HiddenField name='id' value={realEstate.id} />
        : null
    const minValuationDate =
      purchase && purchase.date
        ? moment(purchase.date, DateStorageFormat).add(1, 'd')
        : undefined
    const maxValuationDate =
      sale && sale.date
        ? moment(sale.date, DateStorageFormat).subtract(1, 'd')
        : undefined

    return (
      <form
        className='real-estate-form form form-aligned'
        onSubmit={this.handleSubmit}
      >
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField
            name='name'
            type='text'
            label='Name'
            value={realEstate.name}
            errors={errors.name}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <AddressField
            name='address'
            address={realEstate.address}
            errors={errors.address}
            onChange={this.handleAddressChange}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <TextAreaField
            name='notes'
            label='Notes'
            value={realEstate.notes}
            onChange={value => this.handleChange('notes', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Purchase</legend>
          <DateField
            name='purchaseDate'
            label='Purchase date'
            value={purchase ? purchase.date : undefined}
            errors={errors.purchaseDate}
            forceErrorDisplay={allErrorsShown}
            maxDate={maxValuationDate}
            onChange={value => this.handlePurchaseChange({ date: value })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='purchaseAmount'
            label='Purchase amount'
            value={
              purchase && purchase.amount
                ? purchase.amount.toString()
                : undefined
            }
            errors={errors.purchaseAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value =>
              this.handlePurchaseChange({ amount: parseInt(value, 10) })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Sale</legend>
          <DateField
            name='saleDate'
            label='Sale date'
            value={sale ? sale.date : undefined}
            errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            minDate={minValuationDate}
            onChange={value => this.handleSaleChange({ date: value })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='saleAmount'
            label='Sale amount'
            value={sale && sale.amount ? sale.amount.toString() : undefined}
            errors={errors.saleAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value =>
              this.handleSaleChange({ amount: parseInt(value, 10) })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Value</legend>
          <ValuationsInput
            name='valuations'
            valuations={realEstate.valuations}
            minDate={minValuationDate}
            maxDate={maxValuationDate}
            errors={errors.valuations}
            forceErrorDisplay={allErrorsShown}
            onChange={this.handleValuationsChange}
            onFocus={this.handleFocus}
          />
        </fieldset>
        <fieldset>
          <legend>Mortgage</legend>
          <button type='submit' className='button button-primary'>
            Submit
          </button>
        </fieldset>
      </form>
    )
  }

  static validate(realEstate) {
    const name = []
      .concat(Validations.required(realEstate.name))
      .concat(Validations.minLength(realEstate.name, 3))
      .concat(Validations.maxLength(realEstate.name, 100))
    const address = AddressField.validate(realEstate.address)
    const purchaseDate = [].concat(
      Validations.required(realEstate.purchaseDate)
    )
    const valuations = []
      .concat(ValuationsInput.validate(realEstate.valuations))
      .concat(RealEstateForm.purchaseValuationPresent(realEstate))
      .concat(RealEstateForm.valuationsWithinPurchaseAndSale(realEstate))

    return {
      ...RealEstateErrorsDefaults,
      name,
      address,
      purchaseDate,
      valuations,
    }
  }

  static purchaseValuationPresent(realEstate) {
    return Validations.validate(
      realEstate,
      re => re.valuations.some(valuation => valuation.type === 'purchase'),
      'There must be a purchase valuation'
    )
  }

  static valuationsWithinPurchaseAndSale(realEstate) {
    const purchaseVal = realEstate.valuations.find(v => v.type === 'purchase')
    const saleVal = realEstate.valuations.find(v => v.type === 'sale')
    if (!purchaseVal) {
      return []
    }

    return Validations.validate(
      realEstate,
      re =>
        re.valuations.every(
          valuation =>
            [ 'purchase', 'sale' ].includes(valuation.type) ||
            (saleVal
              ? valuation.date > purchaseVal.date &&
                valuation.date < saleVal.date
              : valuation.date > purchaseVal.date)
        ),
      'Valuations must all be between the purchase and sale dates'
    )
  }
}

RealEstateForm.propTypes = {
  realEstate: PropTypes.shape({
    name: PropTypes.string,
    address: PropTypes.shape({
      line1: PropTypes.string,
      line2: PropTypes.string,
      line3: PropTypes.string,
      locality: PropTypes.string,
      state: PropTypes.string,
      postcode: PropTypes.string,
    }),
    notes: PropTypes.string,
    purchaseDate: PropTypes.string,
    saleDate: PropTypes.string,
    valuations: PropTypes.arrayOf(
      PropTypes.shape({
        date: PropTypes.string,
        amount: PropTypes.number,
        note: PropTypes.string,
        type: PropTypes.oneOf([ 'none', 'purchase', 'sale' ]),
      })
    ),
  }),
  onSubmit: PropTypes.func,
}

export default RealEstateForm
