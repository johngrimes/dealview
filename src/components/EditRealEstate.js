import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'

import RealEstateForm from './RealEstateForm.js'
import Breadcrumbs from './Breadcrumbs.js'
import NotFound from './NotFound.js'
import RealEstateActions from '../actions/realEstate.js'

import './styles/EditRealEstate.css'

export class EditRealEstate extends React.Component {
  constructor(props) {
    super(props)
    if (this.props.realEstate.status === 'uninitialised') {
      this.props.dispatch(RealEstateActions.loadRealEstate())
    }

    this.handleSubmit = this.handleSubmit.bind(this)
  }

  breadcrumbs() {
    const { id, realEstate: { object: realEstate } } = this.props
    const realEstateName = typeof realEstate === 'undefined' ? id : realEstate.name
    return [
      { display: 'Portfolio', path: '/portfolio' },
      { display: 'Assets', path: '/portfolio/assets' },
      { display: realEstateName, path: `/portfolio/assets/real-estate/${id}` },
    ]
  }

  handleSubmit(realEstate) {
    this.props.dispatch(RealEstateActions.putRealEstate(realEstate))
    this.props.history.push('/portfolio/assets')
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.realEstate.status === 'uninitialised') {
      this.props.dispatch(RealEstateActions.loadRealEstate())
    }
  }

  render() {
    return typeof this.props.realEstate.object === 'undefined'
      ? <NotFound />
      : <div className='edit-real-estate'>
          <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
          <RealEstateForm realEstate={this.props.realEstate.object} onSubmit={this.handleSubmit} />
        </div>
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    realEstate: {
      status: state.realEstate.status,
      object: state.realEstate.objects[ownProps.id],
    },
  }
}

export default connect(mapStateToProps)(withRouter(EditRealEstate))
