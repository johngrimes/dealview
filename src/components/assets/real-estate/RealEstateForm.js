// @flow

import React from 'react'
import _ from 'lodash'

import InputField from '../../forms/InputField.js'
import HiddenField from '../../forms/HiddenField.js'
import AddressField, { AddressErrorsDefaults } from '../../forms/AddressField.js'
import TextAreaField from '../../forms/TextAreaField.js'
import ValuationsInput from '../../forms/ValuationsInput.js'
import DateField from '../../forms/DateField.js'
import * as Validations from '../../../utils/FormValidation.js'
import { RealEstateDefaults } from '../../../data/assets/realEstate.js'
import type { RealEstate } from '../../../data/assets/realEstate.js'
import type { Address } from '../../../data/commonTypes.js'
import type { AddressErrors } from '../../forms/AddressField.js'
import type { Valuations } from '../../../components/forms/ValuationsInput.js'
import type { FieldErrors } from '../../../utils/FormValidation.js'

import '../../../styles/forms.css'
import '../../../styles/buttons.css'
import '../../../styles/tables.css'
import './RealEstateForm.css'

type RealEstateErrors = {
  name: FieldErrors,
  address: AddressErrors,
  notes: FieldErrors,
  purchaseDate: FieldErrors,
  saleDate: FieldErrors
}
const RealEstateErrorsDefaults = {
  name: [],
  address: AddressErrorsDefaults,
  notes: [],
  purchaseDate: [],
  saleDate: []
}

type Props = {
  realEstate?: RealEstate,
  onSubmit?: (realEstate: RealEstate) => void
}

type State = {
  realEstate: RealEstate,
  errors?: RealEstateErrors,
  allErrorsShown: boolean,
  focusedInput: string
}

class RealEstateForm extends React.Component {
  props: Props
  state: State
  handleChange: (fieldName: string, value: string) => void
  handleFocus: (fieldName: string) => void
  handleAddressChange: (address: Address) => void
  handleValuationsChange: (valuations: Valuations) => void
  handleSubmit: (event: Event) => boolean

  constructor(props: Props) {
    super(props)
    this.state = {
      realEstate: RealEstateDefaults,
      allErrorsShown: false,
      focusedInput: 'name'
    }

    if (props.realEstate) { this.state.realEstate = props.realEstate }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleAddressChange = this.handleAddressChange.bind(this)
    this.handleValuationsChange = this.handleValuationsChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName: string, value: string): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, [fieldName]: value }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate)
      }
    })
  }

  handleFocus(fieldName: string): void {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleAddressChange(address: Address): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, address }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate)
      }
    })
  }

  handleValuationsChange(valuations: Valuations): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, valuations }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate)
      }
    })
  }

  handleSubmit(event: Event): boolean {
    event.preventDefault()
    this.setState(
      () => ({ errors: RealEstateForm.validate(this.state.realEstate) }),
      () => {
        const { errors, realEstate } = this.state
        if (errors !== undefined && Validations.areErrorsPresent(errors)) {
          const firstErrorFieldName = RealEstateForm.findFirstErrorFieldName(errors)
          firstErrorFieldName
            ? this.setState(() => ({ allErrorsShown: true, focusedInput: firstErrorFieldName }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) { this.props.onSubmit(realEstate) }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.realEstate) {
      this.setState(() => ({ realEstate: nextProps.realEstate }))
    }
  }

  render() {
    const { realEstate, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField = typeof realEstate.id === 'string'
      ? <HiddenField name='id' value={realEstate.id} />
      : null

    return (
      <form className='real-estate-form form form-aligned' onSubmit={this.handleSubmit}>
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField name='name' type='text' label='Name' value={realEstate.name}
            errors={errors.name} forceErrorDisplay={allErrorsShown}
            onChange={(value) => this.handleChange('name', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <AddressField name='address' address={realEstate.address} errors={errors.address}
            onChange={this.handleAddressChange}
            onFocus={this.handleFocus} focus={focusedInput} />
          <TextAreaField name='notes' label='Notes' value={realEstate.notes}
            onChange={(value) => this.handleChange('notes', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Purchase and sale</legend>
          <DateField name='purchaseDate' label='Purchase date'
            value={realEstate.purchaseDate} errors={errors.purchaseDate}
            forceErrorDisplay={allErrorsShown}
            onChange={(value) => this.handleChange('purchaseDate', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <DateField name='saleDate' label='Sale date'
            value={realEstate.saleDate} errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            onChange={(value) => this.handleChange('saleDate', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Value</legend>
          <ValuationsInput name='valuations' valuations={realEstate.valuations}
            onChange={this.handleValuationsChange} onFocus={this.handleFocus} />
        </fieldset>
        <fieldset>
          <legend>Mortgage</legend>
          <button type='submit' className='button button-primary'>Submit</button>
        </fieldset>
      </form>
    )
  }

  static validate(realEstate: RealEstate): RealEstateErrors {
    const errors = RealEstateErrorsDefaults

    errors.name = []
      .concat(Validations.required(realEstate.name))
      .concat(Validations.minLength(realEstate.name, 3))
      .concat(Validations.maxLength(realEstate.name, 100))

    errors.address = AddressField.validate(realEstate.address)

    errors.notes = []

    errors.purchaseDate = []
      .concat(Validations.required(realEstate.purchaseDate))

    return errors
  }

  static findFirstErrorFieldName(realEstateErrors: RealEstateErrors): string|null {
    return (function seek(realEstateErrors, parentComponentName) {
      let match = null
      _.some(realEstateErrors, (errors, fieldName) => {
        if (errors instanceof Array) {
          if (errors.length > 0) {
            match = parentComponentName ? parentComponentName + '-' + fieldName : fieldName
            return true
          } else return false
        } else if (errors instanceof Object) {
          match = seek(errors, fieldName)
          return (match !== null)
        }
      })
      return match
    })(realEstateErrors)
  }
}

export default RealEstateForm
