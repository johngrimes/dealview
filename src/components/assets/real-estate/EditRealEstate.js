// @flow

import React from 'react'
import { connect } from 'react-redux'
import type { Dispatch } from 'redux'

import RealEstateForm from './RealEstateForm.js'
import Breadcrumbs from '../../Breadcrumbs/Breadcrumbs.js'
import NotFound from '../../NotFound/NotFound.js'
import { putRealEstate, loadRealEstate } from '../../../actions/realEstate.js'
import type { RealEstate } from '../../../data/assets/realEstate.js'
import type { GlobalState } from '../../../store.js'
import type { ObjectStoreStatus } from 'store'
import type { BreadcrumbTrail } from '../../../components/Breadcrumbs/Breadcrumbs.js'

import './EditRealEstate.css'

type Props = {
  id: string,
  realEstate: {
    status: ObjectStoreStatus,
    object: RealEstate
  },
  dispatch: Dispatch
}

export class EditRealEstate extends React.Component {
  props: Props
  handleSubmit: (realEstate: RealEstate) => void

  constructor(props: Props) {
    super(props)
    if (this.props.realEstate.status === 'uninitialised') {
      this.props.dispatch(loadRealEstate())
    }

    this.handleSubmit = this.handleSubmit.bind(this)
  }

  breadcrumbs(): BreadcrumbTrail {
    const { id, realEstate: { object: realEstate } } = this.props
    const realEstateName = (typeof realEstate === 'undefined') ? id : realEstate.name
    return [
      { display: 'Portfolio', path: '/portfolio' },
      { display: 'Assets', path: '/portfolio/assets' },
      { display: realEstateName, path: `/portfolio/assets/real-estate/${id}` },
    ]
  }

  handleSubmit(realEstate: RealEstate): void {
    this.props.dispatch(putRealEstate(realEstate))
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.realEstate.status === 'uninitialised') {
      this.props.dispatch(loadRealEstate())
    }
  }

  render() {
    return (typeof this.props.realEstate.object === 'undefined')
    ? <NotFound />
    : <div className='edit-real-estate'>
        <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
        <RealEstateForm realEstate={this.props.realEstate.object} onSubmit={this.handleSubmit} />
      </div>
  }
}

const mapStateToProps = (state: GlobalState, ownProps: Props) => {
  return {
    realEstate: {
      status: state.realEstate.status,
      object: state.realEstate.objects[ownProps.id],
    },
  }
}

export default connect(mapStateToProps)(EditRealEstate)
