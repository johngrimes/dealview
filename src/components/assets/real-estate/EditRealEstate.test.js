/* global expect */

import { EditRealEstate } from './EditRealEstate.js'

import React from 'react'
import { shallow } from 'enzyme'

import RealEstateForm from './RealEstateForm.js'
import { validRealEstateWithId1 } from '../../../test/fixtures/realEstate.js'

describe('EditRealEstate', () => {
  it('should render an existing Real Estate asset', () => {
    const props = {
      id: '73',
      realEstate: {
        status: 'loaded',
        object: validRealEstateWithId1
      }
    }
    const wrapper = shallow(<EditRealEstate {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should pass Real Estate object to RealEstateForm component', () => {
    const props = {
      id: '73',
      realEstate: {
        status: 'loaded',
        object: validRealEstateWithId1
      }
    }
    const wrapper = shallow(<EditRealEstate {...props} />)
    expect(wrapper.find(RealEstateForm).prop('realEstate')).toBe(validRealEstateWithId1)
  })

  it('should load Real Estate if uninitialised', () => {
    const dispatch = jest.fn()
    const realEstateActions = require('../../../actions/realEstate.js')
    realEstateActions.loadRealEstate = jest.fn()
    const props = {
      id: '73',
      realEstate: {
        status: 'uninitialised'
      },
      dispatch
    }
    const wrapper = shallow(<EditRealEstate {...props} />)
    expect(dispatch).toHaveBeenCalled()
    expect(realEstateActions.loadRealEstate).toHaveBeenCalled()
  })

  it('should save Real Estate when handleSubmit is called', () => {
    const dispatch = jest.fn()
    const props = {
      id: '73',
      realEstate: {
        status: 'loaded',
        object: validRealEstateWithId1
      },
      dispatch
    }
    const realEstateActions = require('../../../actions/realEstate.js')
    realEstateActions.putRealEstate = jest.fn()
    const wrapper = shallow(<EditRealEstate {...props} />)
    const handleSubmit = wrapper.find(RealEstateForm).prop('onSubmit')
    handleSubmit(validRealEstateWithId1)
    expect(dispatch).toHaveBeenCalled()
    expect(realEstateActions.putRealEstate).toHaveBeenCalledWith(validRealEstateWithId1)
  })
})
