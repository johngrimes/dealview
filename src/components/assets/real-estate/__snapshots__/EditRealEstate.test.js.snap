// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRealEstate should render an existing Real Estate asset 1`] = `
<div
  className="edit-real-estate"
>
  <Breadcrumbs
    breadcrumbs={
      Array [
        Object {
          "display": "Portfolio",
          "path": "/portfolio",
        },
        Object {
          "display": "Assets",
          "path": "/portfolio/assets",
        },
        Object {
          "display": "5 Sunrise St",
          "path": "/portfolio/assets/real-estate/73",
        },
      ]
    }
  />
  <RealEstateForm
    onSubmit={[Function]}
    realEstate={
      Object {
        "address": Object {
          "line1": "5 Sunrise St",
          "line2": "",
          "line3": "",
          "locality": "Ashgrove",
          "postcode": "4060",
          "state": "QLD",
        },
        "id": "73",
        "name": "5 Sunrise St",
        "notes": "Principal place of residence",
        "purchaseDate": "2014-09-10",
        "valuations": Array [
          Object {
            "amount": 520000,
            "date": "2015-08-25",
            "note": "Purchase price",
          },
          Object {
            "amount": 705000,
            "date": "2016-04-15",
            "note": "Formal valuation",
          },
        ],
      }
    }
  />
</div>
`;
